### PR TITLE
fltrator : init at 2.3

### PR DESCRIPTION
**Pull request by marius851000**
###### Motivation for this change
I wanted to learn how to make a nix package. I've picked a random ( apparently simple ) open source game from libregamewiki, and make a package.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"` - no package depend on it
- [x] Tested execution of all binary files (usually in `./result/bin/`)

- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/games/fltrator/default.nix [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    make HOME=$out RSC_PATH=$out/fltrator
+    '';
+
+  installPhase = ''
+    make HOME=$out RSC_PATH=$out/fltrator install
+    '';
+
+  meta = with stdenv.lib; {
+    description = "A simple retro style arcade side-scroller game.";
+    longDescription = "FLTrator is a simple retro style arcade side-scroller game in which you steer a spaceship through a landscape with hostile rockets and other obstacles.
```
Comment (srhb) [inline pkgs/games/fltrator/default.nix]: Use a multiline string here (`''foo''`) such that leading whitespace gets properly stripped.

File: pkgs/games/fltrator/default.nix [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    ./configure
+    '';
+
+  buildPhase = ''
+    make HOME=$out RSC_PATH=$out/fltrator
+    '';
+
+  installPhase = ''
+    make HOME=$out RSC_PATH=$out/fltrator install
+    '';
```
Comment (srhb) [inline pkgs/games/fltrator/default.nix]: Indent these ending quotes at the same level as `installPhase` (ie. delete two spaces)

File: pkgs/games/fltrator/default.nix [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+  configurePhase = ''
+    cd fltrator*
+    sed -i "s%home += \"fltrator/\"%home = \"$out/fltrator/\"%" src/fltrator.cxx
+    sed -i "s%home += \"fltrator/\"%home = \"$out/fltrator/\"%" src/fltrator-landscape.cxx
+    ./configure
+    '';
+
+  buildPhase = ''
+    make HOME=$out RSC_PATH=$out/fltrator
+    '';
```
Comment (srhb) [inline pkgs/games/fltrator/default.nix]: Indent these ending quotes at the same level as installPhase (ie. delete two spaces)

File: pkgs/games/fltrator/default.nix [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+  unpackPhase = ''
+    unzip $src
+    '';
+
+  configurePhase = ''
+    cd fltrator*
+    sed -i "s%home += \"fltrator/\"%home = \"$out/fltrator/\"%" src/fltrator.cxx
+    sed -i "s%home += \"fltrator/\"%home = \"$out/fltrator/\"%" src/fltrator-landscape.cxx
+    ./configure
+    '';
```
Comment (srhb) [inline pkgs/games/fltrator/default.nix]: Indent these ending quotes at the same level as installPhase (ie. delete two spaces)

File: pkgs/games/fltrator/default.nix [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+  buildInputs = [ fltk unzip which libjpeg ];
+
+  unpackPhase = ''
+    unzip $src
+    '';
+
+  configurePhase = ''
+    cd fltrator*
+    sed -i "s%home += \"fltrator/\"%home = \"$out/fltrator/\"%" src/fltrator.cxx
+    sed -i "s%home += \"fltrator/\"%home = \"$out/fltrator/\"%" src/fltrator-landscape.cxx
```
Comment (srhb) [inline pkgs/games/fltrator/default.nix]: `substituteInPlace` might be slightly more ergonomic here: https://nixos.org/nixpkgs/manual/#ssec-stdenv-functions

Also, this does not make it try and write to `$out` at runtime, right?

File: pkgs/games/fltrator/default.nix [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    url = "https://sourceforge.net/projects/fltrator/files/v${version}/fltrator-${version}-code.zip/download";
+    sha256 = "125aqq1sfrm0c9cm6gyylwdmc8xrb0rjf563xvw7q28sdbl6ayp7";
+  };
+
+
+  buildInputs = [ fltk unzip which libjpeg ];
+
+  unpackPhase = ''
+    unzip $src
+    '';
```
Comment (srhb) [inline pkgs/games/fltrator/default.nix]: Indent these ending quotes at the same level as installPhase (ie. delete two spaces)

File: pkgs/games/fltrator/default.nix [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+  installPhase = ''
+    make HOME=$out RSC_PATH=$out/fltrator install
+    '';
+
+  meta = with stdenv.lib; {
+    description = "A simple retro style arcade side-scroller game.";
+    longDescription = "FLTrator is a simple retro style arcade side-scroller game in which you steer a spaceship through a landscape with hostile rockets and other obstacles.
+    It has ten different levels and a level editor to create new levels or modify the existing."; # from https://libregamewiki.org/FLTrator
+    homepage = http://fltrator.sourceforge.net/;
+    meta.platforms = stdenv.lib.platforms.linux;
```
Comment (srhb) [inline pkgs/games/fltrator/default.nix]: Just `platforms` -- we're already in the `meta` attrset.

**Review by srhb — CHANGES_REQUESTED** [review]
Welcome and thank you for contributing to nixpkgs! :D

This looks good and appears to work just fine. I have left stylistic comments that I'd like you to resolve.
I also think you can just squash the patch commit into the init commit. :)

File: pkgs/games/fltrator/default.nix (hunk 5 above)
Comment (marius851000) [reply pkgs/games/fltrator/default.nix]: config are saved in an fltk-dependant path, but the level editor save modified level in the out path. I don't know how to solve this. I suppose I can left this in this state.

Comment (srhb) [reply pkgs/games/fltrator/default.nix]: Alright, perhaps that can come later. :) Normally that path will be read-only, and some patching may be required to cause the game to save it in the user's homedir or something like that.

File: pkgs/games/fltrator/default.nix [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+  };
+
+
+  buildInputs = [ fltk unzip which libjpeg ];
+
+  unpackPhase = ''
+  unzip $src
+  '';
+
+  configurePhase = ''
```
Comment (srhb) [inline pkgs/games/fltrator/default.nix]: I think I was unclear with these comments. What I meant was that it should look like this:
```
  configurePhase = ''
    cd fltrator*
    substituteInPlace src/fltrator.cxx\
      --replace 'home += "fltrator/"' "home = \"$out/fltrator/\""
    substituteInPlace src/fltrator-landscape.cxx\
      --replace 'home += "fltrator/"' "home = \"$out/fltrator/\""
    ./configure
  '';
```

That is, the contents of the strings is indented two spaces, but the ending quote is at the same level as the attribute (here, configurePhase)

**Review by jtojnar — CHANGES_REQUESTED** [review]

File: pkgs/games/fltrator/default.nix [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+  src = fetchurl {
+    url = "https://sourceforge.net/projects/fltrator/files/v${version}/fltrator-${version}-code.zip/download";
+    sha256 = "125aqq1sfrm0c9cm6gyylwdmc8xrb0rjf563xvw7q28sdbl6ayp7";
+  };
+
+
+  buildInputs = [ fltk unzip which libjpeg ];
+
+  unpackPhase = ''
+  unzip $src
```
[the remark this answers is not in the document]
Comment (marius851000) [reply pkgs/games/fltrator/default.nix]: If I doesn't do that, it say

> unpacking source archive /nix/store/7bzpdcazhyr4ncgii6kgcmk5f0p8dmd8-download
> do not know how to unpack source archive /nix/store/7bzpdcazhyr4ncgii6kgcmk5f0p8dmd8-download

[the remark this answers is not in the document]
Comment (jtojnar) [reply pkgs/games/fltrator/default.nix]: The setup hook checks the file extension: https://github.com/NixOS/nixpkgs/blob/daf6e558013bdc0068e1fbdbe8d1dcfbed2d7cff/pkgs/tools/archivers/unzip/setup-hook.sh#L3. If you change the url as suggested, it should work.

File: pkgs/games/fltrator/default.nix [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+stdenv.mkDerivation rec {
+  name = "fltrator-${version}";
+  version = "2.3";
+
+  src = fetchurl {
+    url = "mirror://sourceforge/fltrator/fltrator-${version}-code.zip";
+    sha256 = "125aqq1sfrm0c9cm6gyylwdmc8xrb0rjf563xvw7q28sdbl6ayp7";
+  };
+
+  buildInputs = [ fltk unzip which libjpeg ];
```
Comment (jtojnar) [inline pkgs/games/fltrator/default.nix]: Dependencies that are run during build should go to `nativeBuildInputs ` (unzip, which)

File: pkgs/games/fltrator/default.nix [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+  '';
+
+  dontAddPrefix = true;
+
+  makeFlags = [ "HOME=$(out)" "RSC_PATH=$(out)/fltrator"];
+
+  postInstall = ''
+    mkdir -p $out/share/applications
+    cp rsc/fltrator.desktop $out/share/applications
+    cp rsc/fltrator-128.png $out/fltrator/icon.png
```
Comment (jtojnar) [inline pkgs/games/fltrator/default.nix]: Icons should go to `$out/share/icons/hicolor/128x128/apps` as per [icon-theme-spec](https://specifications.freedesktop.org/icon-theme-spec/icon-theme-spec-latest.html).

**Comment by jtojnar** [comment]
I would try

```
git reset --hard master
curl -L https://patch-diff.githubusercontent.com/raw/NixOS/nixpkgs/pull/52021.diff | git apply
```

Then add, commit, and force push.

**Comment by aanderse** [comment]
@marius851000 ping

**Comment by marius851000** [comment]
It appear to be good for me.

**Pull request merged**